libraries/AP_Networking/AP_Networking_PPP.cpp

```diff
@@ -0,0 +1,197 @@
+    const uint8_t *ptr = (const uint8_t *)data;
+    while (remaining > 0) {
+        auto n = driver.uart->write(ptr, remaining);
+        if (n > 0) {
+            remaining -= n;
+            ptr += n;
+        } else {
+            hal.scheduler->delay_microseconds(100);
+        }
+    }
+    return len;
+}
+
+/*
+  callback on link status change
+ */
+void AP_Networking_PPP::ppp_status_callback(struct ppp_pcb_s *pcb, int code, void *ctx)
+{
+    auto &driver = *(AP_Networking_PPP *)ctx;
+    struct netif *pppif = ppp_netif(pcb);
+    const char *msg = nullptr;
+
+    switch (code) {
+    case PPPERR_NONE: {
+        // got addresses
+        driver.activeSettings.ip = ntohl(netif_ip4_addr(pppif)->addr);
+        driver.activeSettings.gw = ntohl(netif_ip4_gw(pppif)->addr);
+        driver.activeSettings.nm = ntohl(netif_ip4_netmask(pppif)->addr);
+        driver.activeSettings.last_change_ms = AP_HAL::millis();
+
+        break;
+    }
+    case PPPERR_PARAM: {           /* Invalid parameter. */
+        msg = "PPPERR_PARAM";
+        break;
+    }
+    case PPPERR_OPEN: {            /* Unable to open PPP session. */
+        msg = "PPPERR_OPEN";
+        break;
+    }
+    case PPPERR_DEVICE: {          /* Invalid I/O device for PPP. */
+        msg = "PPPERR_DEVICE";
+        break;
+    }
+    case PPPERR_ALLOC: {           /* Unable to allocate resources. */
+        msg = "PPPERR_ALLOC";
+        break;
+    }
+    case PPPERR_USER: {            /* User interrupt. */
+        msg = "PPPERR_USER";
+        break;
+    }
+    case PPPERR_CONNECT: {         /* Connection lost. */
+        msg = "PPPERR_CONNECT";
+        break;
+    }
+    case PPPERR_AUTHFAIL: {        /* Failed authentication challenge. */
+        msg = "PPPERR_AUTHFAIL";
+        break;
+    }
+    case PPPERR_PROTOCOL: {        /* Failed to meet protocol. */
+        msg = "PPPERR_PROTOCOL";
+        break;
+    }
+    case PPPERR_PEERDEAD: {        /* Connection timeout */
+        msg = "PPPERR_PEERDEAD";
+        break;
+    }
+    case PPPERR_IDLETIMEOUT: {     /* Idle Timeout */
+        msg = "PPPERR_IDLETIMEOUT";
+        break;
+    }
+    case PPPERR_CONNECTTIME: {     /* Max connect time reached */
+        msg = "PPPERR_CONNECTTIME";
+        break;
+    }
+    case PPPERR_LOOPBACK: {        /* Loopback detected */
+        msg = "PPPERR_LOOPBACK";
+        break;
+    }
+    }
+    if (msg != nullptr) {
+        GCS_SEND_TEXT(MAV_SEVERITY_INFO, "PPP: %s", msg);
+    }
+}
+
+
+/*
+  initialise PPP network backend using LWIP
+ */
+bool AP_Networking_PPP::init()
+{
+    auto &sm = AP::serialmanager();
+    uart = sm.find_serial(AP_SerialManager::SerialProtocol_PPP, 0);
+    if (uart == nullptr) {
+        return false;
+    }
+
+    pppif = new netif;
+    if (pppif == nullptr) {
+        return false;
+    }
+
+    // initialise TCP/IP thread
+    LWIP_TCPIP_LOCK();
+    tcpip_init(NULL, NULL);
+    LWIP_TCPIP_UNLOCK();
+
+    hal.scheduler->delay(100);
```

Comment:
why 100?

Reply:
just needs enough time for the tcpip thread to start, could be more elegant, but this works